aignostic/aignostic/dataset/validate_dataset_api.py

```
@@ -1,65 +1,63 @@
+            raise ValueError("All label rows must have the same number of elements.")
 
+    return data
 
-@app.get('/validate-dataset')
-async def validate_dataset(url: str = Query(..., description="Dataset URL")):
-    """
-    Validate a dataset URL and return the columns and number of rows.
+
+@app.post("/fetch-data")
+def fetch_dataset(request: FetchDatasetRequest) -> ModelInput:
     """
-    if not is_valid_url(url):
-        raise HTTPException(status_code=400, detail="Invalid URL")
+    Validate a dataset URL and parse the data to be used in the model.
 
+    Params:
+    - request : FetchDatasetRequest - Pydantic model for the request
+    """
     try:
-        data = fetch_data(url)
-        dataset = parse_dataset(data)
-        return JSONResponse(content={
-            "message": "Data successfully parsed into DataFrame",
-            "columns": dataset.column_names,
-            "rows": len(dataset.rows)
-        })
+        headers = {"Authorization": f"Bearer {request.dataset_api_key}"} if request.dataset_api_key else {}
+        response = requests.get(request.dataset_url, headers=headers)
+
+        response.raise_for_status()
 
-    # TODO: Add logging to exception handling
+        data = response.json()
+        return validate_dataset_format(data)
     except requests.exceptions.RequestException as e:
-        if isinstance(e, requests.exceptions.HTTPError) and e.response.status_code == 404:
-            raise HTTPException(status_code=404, detail="Resource not found")
-        else:
-            raise HTTPException(status_code=400, detail="Failed to fetch data")
-    except Exception:
-        raise HTTPException(status_code=400, detail="Invalid data format")
+        if response.status_code == 401:
+            raise HTTPException(status_code=401, detail="Unauthorized access: Please check your API Key")
+        elif response.status_code == 404:
+            raise HTTPException(status_code=404, detail="Data not found")
```

I don't think we should rewrite any of the exception logic (or catch it) -> for me it makes sense to make 4XX error handling the responsibility of the client and we can propagate the error straight to the frontend without modifying it. This will likely make the errors easier for the client to understand as it's their own error message.

---

Makes sense, have changed